### PR TITLE
Templated SodiumProperties

### LINKED ISSUE #14798: Template SodiumProperties
## Reason
Sodium properties does a nice job combining all sodium properties into one file. In order to use these for AD and non-AD methods, they need to be templated to accept ADReal or Real.

## Design
Template SodiumProperties

## Impact
Usability improvements.

### PR DESCRIPTION
**Pull request by tophmatthews**
Templated SodiumProperties to allow use with ADReal's. Added tests. General cleanup.

Closes #14798

### CODE REVIEW

**Review by andrsd — COMMENTED** [review]
How is your test different than what is in the unit test `unit/src/SodiumPropertiesTest.C`?

**Comment by tophmatthews** [comment]
The biggest change is that `unit/src/SodiumPropertiesTest.C` only tests the values in `SodiumProperties` at a single temperature. The test file here tests at a swath of temperatures. Yes, these are simple polynomials without complex regime changes, but in general it's best to test across more than one temperatures to tease out details.

Also, this test consolidates everything into the place it is expected to be (i.e. in the tests folder in the same module as the UserObject), shows an example of how to utilize the `UserObject` (e.g. `SodiumPropertiesTestMaterial`) and how to set up an input file using the `SodiumProperties` user object.

**Comment by andrsd** [comment]
The preferred way of testing fluid properties user objects is via unit tests. We started with CSV/Exodiff, but as you can see the overhead of building such a test is quite big. The unit test approach is much simpler. If you wish to test more values, then do it in the unit test.

> Also, this test consolidates everything into the place it is expected to be (i.e. in the tests folder in the same module as the UserObject)

That's only because modules cannot have their own unit tests. For example, applications can and our other fluid properties that are not in MOOSE modules are using that approach. Yes, it would be nice to have module-specific unit tests and when we do, the unit tests would be moved here.


> shows an example of how to utilize the `UserObject` (e.g. `SodiumPropertiesTestMaterial`) and how to set up an input file using the `SodiumProperties` user object.

If you want people to use properties from `SodiumProperties` via a material, then do not do it as a test object, because what they are gonna do is they will copy and paste your code into where ever they decide. Note that apps do not build module test objects by default. And we don't want that.

I would suggest you move that material from `test/` into the module, drop the `Test` part, so that it is a regular material and people can just use it - similar to what we do with `FluidPropertiesMaterial`.

**Comment by tophmatthews** [comment]
> The unit test approach is much simpler.

It is simpler for one temperature. It becomes more complicated as you test more temperatures. Sounds like you've already walked both paths and choose unit tests, but I have a hard time thinking the overhead is that much greater due to all the advantages I listed.

**Comment by tophmatthews** [comment]
> If you want people to use properties from SodiumProperties via a material, then do not do it as a test object, because what they are gonna do is they will copy and paste your code into where ever they decide. Note that apps do not build module test objects by default. And we don't want that.

That is not the point of the test object, perhaps I wasn't so clear. Users should use the UserObject method of getting the properties, just like `SodiumPropertiesTestMaterial` does, and just like the test input in this PR. The idea of `SodiumPropertiesTestMaterial` is it shows a clear way to utilize the `UserObject` in their own class, e.g. `_sodium(getUserObject<SodiumProperties>("fp"))`, `_k[_qp] = _sodium.k(_temperature[_qp]);`, etc.

If you think that this PR really detracts more than it adds from the overall design of the `FluidProperties` module, then I can remove `SodiumPropertiesTestMaterial` and add to the unit test, but know that I'd be doing it be-grudgingly :)

**Comment by andrsd** [comment]
Let's start with the material object first. What I am trying to tell you is that the applicability of your test object is much larger and therefore you can make it into an object that people would have to implement is they wanted to use `SodiumProperties` user object and store the properties in material properties. Go look at `FluidPropertiesMaterial`, it does the exact thing like your test material, but with a different API - it uses `p` and `T` as inputs. We have similar materials in RELAP-7 and Sockeye - they also take a fluid properties user object and they store the fluid properties in material properties, but in that case they go via `e` and `v`.  This sodium UO is kind of one-off thing and has its own API, so we cannot use `FluidPropertiesMaterial` for that - it needs a new one - exactly the one you implemented for your test.

Bottom line - we want this material, but not as a test object.

**Comment by andrsd** [comment]
As for the unit test - rough pseudo code:
```
Real k_fn(Real T)
{
  return 124.67 - 0.11381 * T + 5.5226e-5 * std::pow(T, 2) - 1.1842e-8 * std::pow(T, 3);
}

for (Real T = 10; T < 1000; T = T + 100)
{
  REL_TEST(_fp->k(T), k_fn(T), 1.0E-5);
  // similarly for the other quantities
}
```
Adjust the limits and the step of the test range like you want. This looks much simpler to me...

**Comment by cpgr** [comment]
Hey @andrsd - there are some modules with unit tests within the modules that get run for `modules-debug` at least aren't there , see https://civet.inl.gov/job/471159/ for example?

**Comment by andrsd** [comment]
> Hey @andrsd - there are some modules with unit tests within the modules that get run for `modules-debug` at least aren't there , see https://civet.inl.gov/job/471159/ for example?

Yeah, you are right. I must have missed that. In that case, we should move all our unit tests here, because that's were they belong. I noticed some porous flow specific ones there, so you may want to move those as well...

**Comment by andrsd** [comment]
#14808 moves the fluid properties unit test in the fluid properties module...

**Comment by tophmatthews** [comment]
The unit test makes more sense, good way of doing it by defining it's own function and sampling that function.

I'll move the test material out of test, `SodiumPropertiesTestMaterial` -> `SodiumPropertiesMaterial`. Since the test in here for `SodiumPropertiesMaterial` also tests `SodiumProperties` against hand calculations, the unit test seems unnecessary?

**Comment by cpgr** [comment]
> Yeah, you are right. I must have missed that. In that case, we should move all our unit tests here, because that's were they belong. I noticed some porous flow specific ones there, so you may want to move those as well...

I'll do the same thing for porous flow as you have done in #14808

File: modules/fluid_properties/doc/content/source/materials/SodiumPropertiesMaterial.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
+# Sodium Properties Material
+
+!syntax description /Materials/SodiumPropertiesMaterial
+
+`SodiumPropertiesMaterial` samples the functions availble in [SodiumProperties](SodiumProperties.md)
```
Comment (andrsd) [inline modules/fluid_properties/doc/content/source/materials/SodiumPropertiesMaterial.md]: availble -> available

File: modules/fluid_properties/doc/content/source/materials/SodiumPropertiesMaterial.md [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+and provides them as material properties.
+
+The following materials created for liquid sodium are:
```
Comment (andrsd) [inline modules/fluid_properties/doc/content/source/materials/SodiumPropertiesMaterial.md]: probably "...material properties created..."?

File: modules/fluid_properties/doc/content/source/materials/SodiumPropertiesMaterial.md [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+- Thermal Conductivity: `k(temperature)`
+- Enthalpy: `h(temperature)`
+- Heat capacity: `heatCapacity(temperature)`
```
Comment (andrsd) [inline modules/fluid_properties/doc/content/source/materials/SodiumPropertiesMaterial.md]: If this refers to the actual material property name, then it should be `cp`, `heatCapacity` is the name of the method.

File: modules/fluid_properties/doc/content/source/materials/SodiumPropertiesMaterial.md [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+- Temperature: `temperature(h)`
```
Comment (andrsd) [inline modules/fluid_properties/doc/content/source/materials/SodiumPropertiesMaterial.md]: As in above, if this is the material property name, then it is `t_from_h`, but see below as well.

File: modules/fluid_properties/include/materials/SodiumPropertiesMaterial.h [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+  MaterialProperty<Real> & _k;
+
+  /// Enthalpy
+  MaterialProperty<Real> & _h;
+
+  /// Heat capacity
+  MaterialProperty<Real> & _cp;
+
+  /// Temperature from enthalpy
+  MaterialProperty<Real> & _t_from_h;
```
Comment (andrsd) [inline modules/fluid_properties/include/materials/SodiumPropertiesMaterial.h]: I would suggest to use `T` for temperature to be consistent with the rest of the module. Also, little `t` is usually used for time in MOOSE - just to avoid confusion.

File: modules/fluid_properties/include/materials/SodiumPropertiesMaterial.h [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+  /// Density
+  MaterialProperty<Real> & _rho;
+
+  /// Derivative of density with respect to temperature
+  MaterialProperty<Real> & _drho_dt;
```
Comment (andrsd) [inline modules/fluid_properties/include/materials/SodiumPropertiesMaterial.h]: Here `drho_dT` would be better...

File: modules/fluid_properties/include/materials/SodiumPropertiesMaterial.h [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+  /// Derivative of density with respect to enthalpy
+  MaterialProperty<Real> & _drho_dh;
+
+  // Multicomponent fluid properties UserObject
```
Comment (andrsd) [inline modules/fluid_properties/include/materials/SodiumPropertiesMaterial.h]: Sodium, not Multicomponent...

**Review by andrsd — CHANGES_REQUESTED** [review]

**Comment by andrsd** [comment]
> I'll move the test material out of test, `SodiumPropertiesTestMaterial` -> `SodiumPropertiesMaterial`.

:+1:

> Since the test in here for `SodiumPropertiesMaterial` also tests `SodiumProperties` against hand calculations, the unit test seems unnecessary?

Right, we can use your test for actually testing the material. It is a bit of an overkill, but since you already did all the work, we can keep that as is. But, in general, it is much better to keep tests as small as possible...

File: modules/fluid_properties/doc/content/source/materials/SodiumPropertiesMaterial.md [hunk 8]
```diff
@@ -2,15 +2,15 @@
...
 and provides them as material properties.
 
-The following materials created for liquid sodium are:
+The following material properties created for liquid sodium are:
 
 - Thermal Conductivity: `k(temperature)`
 - Enthalpy: `h(temperature)`
-- Heat capacity: `heatCapacity(temperature)`
-- Temperature: `temperature(h)`
+- Heat capacity: `h(temperature)`
```
Comment (andrsd) [inline modules/fluid_properties/doc/content/source/materials/SodiumPropertiesMaterial.md]: `h` -> `cp`

Comment (tophmatthews) [reply modules/fluid_properties/doc/content/source/materials/SodiumPropertiesMaterial.md]: uggg...of course...

**Review by andrsd — APPROVED** [review]
